Emulator/SPD.File.Emulator/SpdEmulatorApi.cs

```diff
@@ -0,0 +1,135 @@
+        _logger = logger;
+    }
+    public bool TryCreateFromFileSlice(string sourcePath, long offset, string route, string destinationPath)
+    {
+        _logger.Info("[SpdEmulatorApi] TryCreateFromFileSlice: {0}, Ofs {1}, Route {2}", sourcePath, offset, route);
+        var handle = Native.CreateFileW(sourcePath, FileAccess.Read, FileShare.ReadWrite, IntPtr.Zero, FileMode.Open, FileAttributes.Normal, IntPtr.Zero);
+        if (handle == new IntPtr(-1))
+        {
+            _logger.Error("[SpdEmulatorApi] TryCreateFromFileSlice: Failed to open base file with Win32 Error: {0}, Path {1}", Marshal.GetLastWin32Error(), sourcePath);
+            return false;
+        }
+
+        IEmulatedFile? emulated = null;
+        _ = Native.SetFilePointerEx(handle, offset, IntPtr.Zero, 0);
+        if (!_spdEmulator.TryCreateEmulatedFile(handle, sourcePath, destinationPath, route, ref emulated, out _))
+        {
+            _logger.Error("[SpdEmulatorApi] TryCreateFromFileSlice: Failed to Create Emulated File at Path {0}", sourcePath);
+            return false;
+        }
+
+        _logger.Info("[SpdEmulatorApi] TryCreateFromFileSlice: Registering {0}", destinationPath);
+        _framework.RegisterVirtualFile(destinationPath, emulated!);
+        return true;
+    }
+
+    public RouteGroupTuple[] GetEmulatorInput()
+    {
+        // Map input to API.
+        var input = _spdEmulator.GetInput();
+        var result = GC.AllocateUninitializedArray<RouteGroupTuple>(input.Count);
+        for (int x = 0; x < result.Length; x++)
+        {
+            var original = input[x];
+            result[x] = new RouteGroupTuple()
+            {
+                Route = original.Route.FullPath,
+                Files = new DirectoryFilesGroup()
+                {
+                    Files = original.Files.Files,
+                    Directory = new DirectoryInformation()
+                    {
+                        FullPath = original.Files.Directory.FullPath,
+                        LastWriteTime = original.Files.Directory.LastWriteTime
+                    }
+                }
+            };
+        }
+
+        return result;
+    }
+
+    public void InvalidateFile(string spdPath)
+    {
+        _spdEmulator.UnregisterFile(spdPath);
+        _framework.UnregisterVirtualFile(spdPath);
+    }
+
+    public void RegisterSpd(string sourcePath, string destinationPath)
+    {
+        var handle = Native.CreateFileW(sourcePath, FileAccess.Read, FileShare.ReadWrite, IntPtr.Zero, FileMode.Open, FileAttributes.Normal, IntPtr.Zero);
+        if (handle == new IntPtr(-1))
+        {
+            _logger.Error("[SpdEmulatorApi] RegisterSpd: Failed to open spd file with Win32 Error: {0}, Path {1}", Marshal.GetLastWin32Error(), sourcePath);
+            return;
+        }
+
+        _ = Native.SetFilePointerEx(handle, 0, IntPtr.Zero, 0);
+
+        var fileStream = new FileStream(new SafeFileHandle(handle, false), FileAccess.Read);
```

Comment:
Files already on Disk should not be copied to memory when registered, as it uses more RAM unnecessarily. Use the FileStream Directly.

@AnimatedSwine37 This also affects BF emulator.